catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/security/accesscontrol/AccessControlPreQueryPlugin.java

```diff
@@ -0,0 +1,237 @@
+        .filter(String.class::isInstance)
+        .map(String.class::cast)
+        .collect(Collectors.toSet());
+  }
+
+  // Invoked by listener in blueprint
+  public void bindTag(ServiceReference ref) {
+    if (ref == null) {
+      LOGGER.trace("New reference to metacard type was null");
+      return;
+    }
+    Object tag = ref.getProperty(ACCESS_CONTROLLED_TAG);
+    if (tag == null) {
+      LOGGER.trace("Newly registered metacard type {} is not access-controlled", ref);
+      return;
+    }
+    if (!(tag instanceof String)) {
+      LOGGER.error(
+          "The service property 'access-controlled-tag' is not valid, verify blueprint is correct");
+      return;
+    }
+    accessControlledTags.add((String) tag);
+  }
+
+  // Invoked by listener in blueprint
+  public void unbindTag(ServiceReference ref) {
+    if (ref == null) {
+      LOGGER.trace("Reference to old metacard type was null");
+      return;
+    }
+    Object tag = ref.getProperty(ACCESS_CONTROLLED_TAG);
+    if (tag == null) {
+      LOGGER.trace("Deregistered metacard type {} is not access-controlled", ref);
+      return;
+    }
+    if (!(tag instanceof String)) {
+      LOGGER.error(
+          "The service property 'access-controlled-tag' is not valid, verify blueprint is correct");
+      return;
+    }
+    accessControlledTags.remove(tag);
+  }
+
+  @Override
+  public QueryRequest process(QueryRequest input)
+      throws PluginExecutionException, StopProcessingException {
+    final String subjectIdentifier = getSubjectIdentifier();
+    final Set<String> groups = new HashSet<>(getSubjectGroups());
+    final String groupThatCanSeeEverything = configuration.getSystemUserAttributeValue();
+
+    if (groups.contains(groupThatCanSeeEverything)) {
+      LOGGER.debug(
+          "Will not modify filter; subject ({}) had at least one group [{}] that was exempt [{}]",
+          subjectIdentifier,
+          groups,
+          groupThatCanSeeEverything);
+      return input;
+    }
+
+    final TagAggregationVisitor tagVisitor = new TagAggregationVisitor();
+    final SimplifyingFilterVisitor simpleVisitor = new SimplifyingFilterVisitor();
+
+    final Query query = input.getQuery();
+    final Filter simplified = (Filter) query.accept(simpleVisitor, null);
+
+    simplified.accept(tagVisitor, null);
+
+    final Set<String> discoveredTags = tagVisitor.getTags();
+    if (CollectionUtils.isEmpty(discoveredTags)) {
+      LOGGER.debug(
+          "Will not modify filter; subject's ({}) query [{}] did not completely bind results to tags",
+          subjectIdentifier,
+          query);
+      return input;
+    }
+
+    Set<String> tagsNotInCollaborationSet =
+        Sets.filter(discoveredTags, tag -> !accessControlledTags.contains(tag));
+    if (CollectionUtils.isNotEmpty(tagsNotInCollaborationSet)) {
+      LOGGER.debug(
+          "Will not modify filter; subject's ({}) query [{}] referenced tags not in collab set [{}]",
```

Comment:
is anyone going to know what a collaboration set is?

Reply:
Good point. I've changed the variable and the log to use the access controlled verbiage.